src/Collections/Sequences/Seq.dfy

```diff
@@ -0,0 +1,720 @@
+  {
+    a := new T[|s|](i requires 0 <= i < |s| => s[i]);
+  }
+
+  /* converts a sequence to a set */
+  function method {:opaque} ToSet<T>(s: seq<T>): set<T> 
+  {
+    set x: T | x in s
+  }
+
+  /* proves that the cardinality of a subsequence is always less than or equal to that
+  of the full sequence */
+  lemma LemmaCardinalityOfSet<T>(s: seq<T>)
+    ensures |ToSet(s)| <= |s| 
+  {
+    reveal ToSet();
+    if |s| == 0 {
+    } else {
+      assert ToSet(s) == ToSet(DropLast(s)) + {Last(s)};
+      LemmaCardinalityOfSet(DropLast(s));
+    }
+  }
+  
+  /* the cardinality of a subsequence of an empty sequence is also 0 */
+  lemma LemmaCardinalityOfEmptySetIs0<T>(s: seq<T>)
+    ensures |ToSet(s)| == 0 <==> |s| == 0
+  {
+    reveal ToSet();
+    if |s| != 0 {
+      assert s[0] in ToSet(s);
+    }
+  }
+
+  /* is true if there are no duplicate values in the sequence */
+  predicate {:opaque} HasNoDuplicates<T>(s: seq<T>) 
+  {
+    (forall i, j {:trigger s[i], s[j]}:: 0 <= i < |s| && 0 <= j < |s| && i != j ==> s[i] != s[j])
+  }
+
+  /* if sequence a and b don't have duplicates and there are no elements in
+  common between them, then the concatenated sequence of a + b will not contain
+  duplicates either */
+  lemma {:timeLimitMultiplier 3} LemmaNoDuplicatesInConcat<T>(a: seq<T>, b: seq<T>)
+    requires HasNoDuplicates(a);
+    requires HasNoDuplicates(b);
+    requires multiset(a) !! multiset(b);
+    ensures HasNoDuplicates(a+b);
+  {
+    reveal HasNoDuplicates();
+    var c := a + b;
+    if |c| > 1 {
+      assert forall i, j {:trigger c[i], c[j]}:: i != j && 0 <= i < |a| && |a| <= j < |c| ==>
+        c[i] in multiset(a) && c[j] in multiset(b) && c[i] != c[j]; 
+    }
+  }
+
+  /* A sequence with no duplicates converts to a set of the same cardinality */
+  lemma LemmaCardinalityOfSetNoDuplicates<T>(s: seq<T>)
+    requires HasNoDuplicates(s)
+    ensures |ToSet(s)| == |s|
+  {
+    reveal HasNoDuplicates();
+    reveal ToSet();
+    if |s| == 0 {
+    } else {
+      LemmaCardinalityOfSetNoDuplicates(DropLast(s));
+      assert ToSet(s) == ToSet(DropLast(s)) + {Last(s)};
+    }
+  }
+
+  /* proves that there are no duplicate values in the multiset version of the sequence */
+  lemma LemmaMultisetHasNoDuplicates<T>(s: seq<T>)
+    requires HasNoDuplicates(s)
+    ensures forall x {:trigger multiset(s)[x]} | x in multiset(s):: multiset(s)[x] == 1
+  {
+    if |s| == 0 {
+    } else {
+      assert s == DropLast(s) + [Last(s)];
+      assert Last(s) !in DropLast(s) by {
+        reveal HasNoDuplicates();
+      }
+      assert HasNoDuplicates(DropLast(s)) by {
+        reveal HasNoDuplicates();
+      }
+      LemmaMultisetHasNoDuplicates(DropLast(s));
+    }
+  }
+
+  /* finds the index of a certain element in the sequence */
+  function method {:opaque} IndexOf<T(==)>(s: seq<T>, v: T): (i: nat)
+    requires v in s
+    ensures i < |s| && s[i] == v
+  {
+    if s[|s|-1] == v then |s| - 1 else IndexOf(s[..|s|-1], v)
```

Comment:
Just curious why this indexes from the back with `s[|s|-1]` instead of the front with `s[0]`.

Reply:
I did this to match `IndexOfOption()`, where indexing from the back gives us the index immediately when we find `v`, so we can directly return `Some(|s| - 1)`.

Reply:
Ah, makes sense, thanks for pointing this out! I renamed this method `LastIndexOf` and added a separate `IndexOf` method that finds the first occurrence of an element.